### ISSUE TITLE
Certain entities going unavailable after a single 401 and engineOilTemp disappeared from API completely

### ISSUE THREAD

**Issue by mojomojoman** [itchannel/fordpass-ha#364]
Not sure if this is two issues or one but late last night the 'engineOilTemp' data seems to have disappeared from the API which, since I have an EV, is certainly not a useful entity but it had existed prior to last night and lots of other useless entities to me still exist like "lowEngineOilPressure" so I dont think Ford is suddenly making an effort to prune unneeded data but who knows.  Then many hours later my Odometer, remote start status, and messages also disappeared. 

![image](https://github.com/itchannel/fordpass-ha/assets/67213346/92aa3db8-991e-40a8-ac0d-8c316b4ec8bd)

The 'engineOilTemp' error started much earlier and looks like this:
```
2023-10-13 22:59:22.945 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 233, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 389, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 172, in async_update_listeners
    update_callback()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 469, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 779, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 879, in _async_write_ha_state
    state, attr = self._async_generate_attributes()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 820, in _async_generate_attributes
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 785, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/config/custom_components/fordpass/sensor.py", line 574, in state
    return self.get_value("state")
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fordpass/sensor.py", line 222, in get_value
    return self.data["engineOilTemp"]["value"]
           ~~~~~~~~~^^^^^^^^^^^^^^^^^
KeyError: 'engineOilTemp'
```

Oddly it was not until many hours later (~8AM) that the entity's data actually went to Unavailable despite that engineOilTemp error starting in the middle of the night (and repeating with every poll since).  When that entity went dark so did the Odometer and others as shown above.

![image](https://github.com/itchannel/fordpass-ha/assets/67213346/8fcd1299-2b8b-4e44-b0f6-e778969cec2b)

There was a single 401 right at the time those entities dropped off and here's those errors:

```
2023-10-14 08:06:24.653 WARNING (MainThread) [custom_components.fordpass] 401 Client Error: Unauthorized for url: https://api.mps.ford.com/api/expdashboard/v1/details/
2023-10-14 08:06:24.653 WARNING (MainThread) [custom_components.fordpass] Error communicating with FordPass for ##VIN_REDACTED##
2023-10-14 08:06:24.654 ERROR (MainThread) [custom_components.fordpass] Error fetching fordpass data: Error communicating with FordPass for ##VIN_REDACTED##
2023-10-14 08:06:24.666 ERROR (MainThread) [homeassistant.components.template.template_entity] TemplateError('ValueError: Template error: round got invalid input 'None' when rendering template '{{ (state_attr('sensor.fordpass_tirepressure', 'FRONT_LEFT')) | round(0)  }}' but no default was specified') while processing template 'Template<template=({{ (state_attr('sensor.fordpass_tirepressure', 'FRONT_LEFT')) | round(0)  }}) renders=6>' for attribute '_attr_native_value' in entity 'sensor.mach_e_tire_left_front'
2023-10-14 08:06:24.667 ERROR (MainThread) [homeassistant.components.template.template_entity] TemplateError('ValueError: Template error: round got invalid input 'None' when rendering template '{{ (state_attr('sensor.fordpass_tirepressure', 'FRONT_RIGHT')) | round(0)  }}' but no default was specified') while processing template 'Template<template=({{ (state_attr('sensor.fordpass_tirepressure', 'FRONT_RIGHT')) | round(0)  }}) renders=6>' for attribute '_attr_native_value' in entity 'sensor.mach_e_tire_right_front'
2023-10-14 08:06:24.667 ERROR (MainThread) [homeassistant.components.template.template_entity] TemplateError('ValueError: Template error: round got invalid input 'None' when rendering template '{{ (state_attr('sensor.fordpass_tirepressure', 'REAR_LEFT')) | round(0)  }}' but no default was specified') while processing template 'Template<template=({{ (state_attr('sensor.fordpass_tirepressure', 'REAR_LEFT')) | round(0)  }}) renders=6>' for attribute '_attr_native_value' in entity 'sensor.mach_e_tire_left_rear'
2023-10-14 08:06:24.668 ERROR (MainThread) [homeassistant.components.template.template_entity] TemplateError('ValueError: Template error: round got invalid input 'None' when rendering template '{{ (state_attr('sensor.fordpass_tirepressure', 'REAR_RIGHT')) | round(0)  }}' but no default was specified') while processing template 'Template<template=({{ (state_attr('sensor.fordpass_tirepressure', 'REAR_RIGHT')) | round(0)  }}) renders=6>' for attribute '_attr_native_value' in entity 'sensor.mach_e_tire_right_rear'
2023-10-14 08:06:24.675 ERROR (MainThread) [homeassistant.helpers.event] Error while dispatching event for sensor.fordpass_elveh to <Job track state_changed event {'sensor.fordpass_elveh'} HassJobType.Callback <bound method TrackTemplateResultInfo._refresh of <TrackTemplateResultInfo {Template<template=({{states('sensor.fordpass_elveh')}}) renders=6>: <RenderInfo Template<template=({{states('sensor.fordpass_elveh')}}) renders=6> all_states=False all_states_lifecycle=False domains=frozenset() domains_lifecycle=frozenset() entities=frozenset({'sensor.fordpass_elveh'}) rate_limit=None has_time=False exception=None is_static=False>, Template<template=(mdi:gauge) renders=4>: <RenderInfo Template<template=(mdi:gauge) renders=4> all_states=False all_states_lifecycle=False domains=frozenset() domains_lifecycle=frozenset() entities=frozenset() rate_limit=None has_time=False exception=None is_static=True>}>>>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 593, in state
    numerical_value = float(value)  # type:ignore[arg-type]
                      ^^^^^^^^^^^^
ValueError: could not convert string to float: 'unavailable'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/event.py", line 293, in _async_dispatch_entity_id_event
    hass.async_run_hass_job(job, event)
  File "/usr/src/homeassistant/homeassistant/core.py", line 636, in async_run_hass_job
    hassjob.target(*args)
  File "/usr/src/homeassistant/homeassistant/helpers/event.py", line 1211, in _refresh
    self.hass.async_run_hass_job(self._job, event, updates)
  File "/usr/src/homeassistant/homeassistant/core.py", line 636, in async_run_hass_job
    hassjob.target(*args)
  File "/usr/src/homeassistant/homeassistant/components/template/template_entity.py", line 429, in _handle_results
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 779, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 879, in _async_write_ha_state
    state, attr = self._async_generate_attributes()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 820, in _async_generate_attributes
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 785, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/sensor/__init__.py", line 595, in state
    raise ValueError(
ValueError: Sensor sensor.mach_e_range has device class 'distance', state class 'None' unit 'mi' and suggested precision 'None' thus indicating it has a numeric value; however, it has the non-numeric value: 'unavailable' (<class 'str'>)
2023-10-14 08:06:24.714 ERROR (MainThread) [homeassistant.components.template.template_entity] TemplateError('ValueError: Template error: float got invalid input 'None' when rendering template '{{ ((float(state_attr('sensor.fordpass_odometer', 'value')) / 1.60934) | round(0)) }}' but no default was specified') while processing template 'Template<template=({{ ((float(state_attr('sensor.fordpass_odometer', 'value')) / 1.60934) | round(0)) }}) renders=6>' for attribute '_attr_native_value' in entity 'sensor.mach_e_odometer'
2023-10-14 08:07:46.132 WARNING (SyncWorker_2) [custom_components.places.sensor] (Mach-E Geolocation) Tracked Entity (device_tracker.fordpass_tracker) Latitude/Longitude is not set or is not a number. Not Proceeding with Update.
2023-10-14 08:09:27.221 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 233, in _handle_refresh_interval
    await self._async_refresh(log_failures=True, scheduled=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 389, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 172, in async_update_listeners
    update_callback()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 469, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 779, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 879, in _async_write_ha_state
    state, attr = self._async_generate_attributes()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 820, in _async_generate_attributes
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 785, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/config/custom_components/fordpass/sensor.py", line 574, in state
    return self.get_value("state")
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/fordpass/sensor.py", line 222, in get_value
    return self.data["engineOilTemp"]["value"]
           ~~~~~~~~~^^^^^^^^^^^^^^^^^
KeyError: 'engineOilTemp'
```
Sorry about the extra noise there from my template sensors in config.yaml. I guess I need to add conditionals to those so they deal better with unexpected data types.

Not sure why those are the only entities affected or if this has any relation to the token issue I saw as open but this is the ONLY 401 error occurrence I could find in the whole log file which starts on 10/13 around noon and ends just a bit ago but ever since then those entities have remained Unavailable.

I grabbed fresh output from Autonomic.py just now and indeed the engineOilTemp has just disappeared but the Odometer data is there.

[mojomojoman_status_2023-10-14_17_14_25.zip](https://github.com/itchannel/fordpass-ha/files/12908188/mojomojoman_status_2023-10-14_17_14_25.zip)

Now that i'm done documenting this I'll try reloading the integration and/or rebooting HA to see if my entities come back... aaand a reload did bring them back. Well, except for engineOilTemp which can rest in peace but after the reload the error has stopped appearing in the HA logs.

**Comment by itchannel** [comment]
I think Ford made some updates which took the API offline for a little while we in return broke part of my code. I've added in error checking in the next version so should handle the API being down for a little while and won't show unavailable on sensors will just show an error in the log and restore the previous sensor data.

I'll also add a check for that oil temp issue.

**Comment by mojomojoman** [comment]
Did the oil temp sensor disappear for you as well or just me?

yeah clearly its a work in progress on Ford's side so we might need to be prepared for more things dropping out or appearing suddenly warranting additional error checking on refreshes.

**Comment by itchannel** [comment]
I've never had the oil temp on my vehicle from the start. Only temperature sensors I get is coolant.

**Comment by SquidBytes** [comment]
I added a check in my recent pull request that did not display an attribute I added (tripFuelEconomy) if you had an EV. 
I was considering doing this for other sensors and/or attributes to "fine tune" unneeded sensors for specific vehicles, but I think itchannels possible solution of only adding sensors if a vehicle supports it is better.

I'm curious on other EV owers' thoughts / perspective on if it matters or not.
